### ISSUE TITLE
Segmentation Fault after accidentally creating a database with an invalid name

### ISSUE THREAD

**Issue by bastianwegge** [pgbackrest/pgbackrest#1117]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.27

2. PostgreSQL version: 12

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
- database server: Ubuntu 20.04 LTS
- repository host server: Ubuntu 18.04.4 LTS

4. Did you install pgBackRest from source or from a package?
From Source

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
    - errors in the postgresql log file before or during the time you experienced the issue
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

7. Describe the issue:

A database with the name "test?pool=5" was accidentally created. After this incident `pgbackrest` keeps exiting `pgbackrest --stanza=warehouse restore` with `Segmentation fault`.

Using `pgbackrest --stanza=warehouse --log-level-console=trace restore` I found out that there might be a problem while casting the string, which is the faulted database name `test?pool=5`, to json. Please take a look at the following result of the trace.

As you might have guessed, we're in the restore scenario and we had to stop working until we are able to remove this line/definition from the manifest. Right now we can't even decrypt the `backup.manifest`. We tried `openssl aes-256-cbc -d -in ./backup.manifest -out backup.manifest.decrypted` to maybe even get a look but the decryption didn't work.

```
...
2020-07-23 19:34:12.488 P00  TRACE:         common/io/read::ioReadLineParam: (this: {IoRead}, allowEof: true)
2020-07-23 19:34:12.488 P00  TRACE:         common/io/read::ioReadLineParam: => {"test={"db-id":11849206,"db-last-system-id":13432}"}
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 1, size: 1, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 13, size: 13, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 1, size: 1, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 44, size: 44, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/type/json::jsonToVar: (json: {"{"db-id":11849206,"db-last-system-id":13432}"})
2020-07-23 19:34:12.488 P00  TRACE:         common/type/json::jsonToVar: => {KeyValue}
2020-07-23 19:34:12.488 P00  TRACE:         common/io/read::ioReadEof: (this: {IoRead})
2020-07-23 19:34:12.488 P00  TRACE:         common/io/read::ioReadEof: => false
2020-07-23 19:34:12.488 P00  TRACE:         common/io/read::ioReadLineParam: (this: {IoRead}, allowEof: true)
2020-07-23 19:34:12.488 P00  TRACE:         common/io/read::ioReadLineParam: => {"test?pool=5={"db-id":8017549,"db-last-system-id":13432}"}
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 1, size: 1, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 18, size: 18, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 1, size: 1, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: (this: {CryptoHash}, message: {used: 45, size: 45, limit: <off>})
2020-07-23 19:34:12.488 P00  TRACE:         common/crypto/hash::cryptoHashProcess: => void
2020-07-23 19:34:12.488 P00  TRACE:         common/type/json::jsonToVar: (json: {"5={"db-id":8017549,"db-last-system-id":13432}"})
2020-07-23 19:34:12.488 P00  TRACE:         common/type/json::jsonToVar: => {5}
Segmentation fault
```

**Comment by dwsteele** [comment]
Unfortunately this is a known issue (#730) which is due to the unfortunate choice of using INI format for our manifests. `=` in database or file names causes a problem and it's hard to fix without breaking the manifest format.

OK -- to address the issue:

1) The encryption passphrase for manifests is not `repo1-cipher-pass`. The manifest passphrase is stored in `backup/<stanza>/backup.info` (which *is* encrypted with `repo1-cipher-pass`). You use the cipher-pass you find in `backup.info` to decrypt `backup.manifest`.
2) Once you have it decrypted you can update the database name. Then encrypt again and try a restore.
3) You'll immediately get a checksum error (because you modified the file) but the error message will tell you what the checksum should have been. Update `backup.manifest` with the correct checksum (you'll find it at the very end of the file) then encrypt it again. This time the restore should work.

Let use know if you run into problems.

**Comment by bastianwegge** [comment]
Thanks for the immediate answer @dwsteele ! I tried to decrypt the `backup.info` file using the `repo1-cipher-pass` but I keep getting `bad decrypt` error messages. I noticed that I typed `repo-cipher-type=aes-256-cbc` instead of `repo1-cipher-type`, but hitting `pgbackrest info` tells me that the repository is encrypted using `aes-256-cbc` as you can see down below. 

Here's what I'm trying exactly: `openssl aes-256-cbc -d -in ./backup.info -out ./backup.info.decrypted`
It's then asking for the password, where I'm providing the password taken from `repo1-cipher-pass`.

This results in:
```
(user@repo:/home/pgbackrest/repo/backup/warehouse)
-> openssl aes-256-cbc -d -in ./backup.info -out ./backup.info.decrypted
enter aes-256-cbc decryption password:
*** WARNING : deprecated key derivation used.
Using -iter or -pbkdf2 would be better.
bad decrypt
139836314431936:error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt:../crypto/evp/evp_enc.c:537:
```

Output of `pgbackrest info`:
```
stanza: warehouse
    status: ok
    cipher: aes-256-cbc

    db (current)
        wal archive min/max (12-1): 00000005000000B00000002F/000000050000010C00000013

        full backup: 20200712-033005F
            timestamp start/stop: 2020-07-12 03:30:05 / 2020-07-12 05:02:07
            wal start/stop: 00000005000000B00000002F / 00000005000000B000000070
            database size: 126.2GB, backup size: 126.2GB
            repository size: 91.0GB, repository backup size: 91.0GB

        full backup: 20200719-033004F
            timestamp start/stop: 2020-07-19 03:30:04 / 2020-07-19 06:23:02
            wal start/stop: 00000005000000EA000000D4 / 00000005000000EB00000044
            database size: 232.8GB, backup size: 232.8GB
            repository size: 170.2GB, repository backup size: 170.2GB
```

**Comment by dwsteele** [comment]
Yeah, openssl wants to use `md5` by default but we use `sha1` for better security. You need to add ` -md sha1` to the openssl command.

**Comment by dwsteele** [comment]
FYI: here's what a working command looks like: `openssl enc -d -k YOURKEY -md sha1 -aes-256-cbc -in archive.info`.

**Comment by bastianwegge** [comment]
@dwsteele thanks for the detailed description, with your response I successfully decrypted the `backup.info` and used the password I got for the `backup.manifest` to decrypt, edit and encrypt it again. Unfortunately the error stays the same. I'm not getting any checksum error messages and the output is the same I posted above. Since I'm executing the `restore` on a remote machine, is there any caching mechanism integrated into pgbackrest that stores like a global manifest? I edited the manifest for both of the latest backups, which both had the `?pool=5` database inside.

**Comment by dwsteele** [comment]
Right. Because the checksum is failing on `backup.manifest` it is failing over to `backup.manifest.copy`. You'll need to overwrite `backup.manifest.copy` as well. *Please be sure to make a backup of the manifest before overwriting both copies!*

**Comment by bastianwegge** [comment]
So if I understood correctly, I should set the checksum at the bottom of the `backup.manifest` to be what it is expected. 

So first:

```
$ pgbackrest --stanza=warehouse restore

ERROR: [026]: unable to load backup manifest file '/home/pgbackrest/repo/backup/warehouse/20200719-033004F/backup.manifest' or '/home/pgbackrest/repo/backup/warehouse/20200719-033004F/backup.manifest.copy':
       ChecksumError: invalid checksum, actual '4da86f8b09760f24477f741d1b9d52a8a4cf878d' but expected '127b1960733e632fe63e192d698e22d74f13cf7b'
       ChecksumError: invalid checksum, actual '4da86f8b09760f24477f741d1b9d52a8a4cf878d' but expected '127b1960733e632fe63e192d698e22d74f13cf7b'

```

I then went into backup.manifest to set the checksum to `127b1960733e632fe63e192d698e22d74f13cf7b` but the next try it's telling me that I should have set it to the "actual" value.

```
$ pgbackrest --stanza=warehouse restore

ERROR: [026]: unable to load backup manifest file '/home/pgbackrest/repo/backup/warehouse/20200719-033004F/backup.manifest' or '/home/pgbackrest/repo/backup/warehouse/20200719-033004F/backup.manifest.copy':
       ChecksumError: invalid checksum, actual '886eda25e07f8f0524856cc133a20e9126bda259' but expected '4da86f8b09760f24477f741d1b9d52a8a4cf878d'
       ChecksumError: invalid checksum, actual '886eda25e07f8f0524856cc133a20e9126bda259' but expected '4da86f8b09760f24477f741d1b9d52a8a4cf878d'
```

Am I blind or missunderstanding something?

**Comment by dwsteele** [comment]
You need to set it to `4da86f8b09760f24477f741d1b9d52a8a4cf878d` which is the *actual* checksum vs the expected checksum stored in the file.

**Comment by dwsteele** [comment]
Though I wonder if you maybe removed/added a linefeed in the file.

**Comment by dwsteele** [comment]
An extra linefeed won't matter, so try `886eda25e07f8f0524856cc133a20e9126bda259`.

BTW -- just have to ask. Is this a production outage or is it possible to rename the database and make a new backup?

**Comment by bastianwegge** [comment]
@dwsteele this unfortunately is a production outage.

**Comment by bastianwegge** [comment]
Using `886eda25e07f8f0524856cc133a20e9126bda259` worked and the restore process seems to have started. Thanks a lot for your help and patience! I was kind of confused by "actually" and "expected".

I didn't remove/add a LF in the file, I changed the name of the accidentally created database to contain no special-characters, changed the checksum in the last line to be what it prompted me to be and that was it.

**Comment by dwsteele** [comment]
Just to be sure to `diff` the original `backup.manifest` and your updated version. Check that the only updated lines are the db entry and the checksum. If any *files* were accidentally removed from the manifest they will not be restored.

Sorry for the bug and we'll bump up the priority on fixing it, but please be sure to test your restores proactively. It can save a lot of trouble in crunch time.

**Comment by bastianwegge** [comment]
The restore went smoothly, we're live again!

Thank you very much @dwsteele for your help! This would have been a pain for me probably keeping me busy over the weekend.

On the topic of testing restores: The database `test?pool=5` was created on last Saturday 1 day before the last full-backup, the last restore-test on our hot-standby ran last friday. Very very unfortunate so to say.

I'm closing this issue, since my case is done and #730 seems to be the bumper issue. I'm wondering if you guys are interested in a tutorial using `pgbackrest` like we are right now. I'd be happy to contribute in any way to give something back.

**Comment by dwsteele** [comment]
> The restore went smoothly, we're live again!

Glad to hear it!

> On the topic of testing restores: The database test?pool=5 was created on last Saturday 1 day before the last full-backup, the last restore-test on our hot-standby ran last friday. Very very unfortunate so to say.

Yes, that's pretty unfortunate timing.

We have a proposed fix for this problem at #1120. You are welcome to have a look and see if it seems sensible.
